libdevcore/DBFactory.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+    This file is part of cpp-ethereum.
+
+    cpp-ethereum is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    cpp-ethereum is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "DBFactory.h"
+#include "FileSystem.h"
+#include "LevelDB.h"
+#include "MemoryDB.h"
+
+namespace dev
+{
+namespace db
+{
+namespace fs = boost::filesystem;
+namespace po = boost::program_options;
+
+auto g_kind = DatabaseKind::LevelDB;
+fs::path g_dbPath = getDataDir();
+
+/// A helper type to build the tabled of DB implementations.
+///
+/// More readable than std::tuple.
+/// Fields are not initialized to allow usage of construction with initializer lists {}.
+struct DBKindTableEntry
+{
+    DatabaseKind kind;
+    const char* name;
+};
+
+/// The table of available DB implementations.
+///
+/// We don't use a map to avoid complex dynamic initialization. This list will never be long,
+/// so linear search only to parse command line arguments is not a problem.
+DBKindTableEntry dbKindsTable[] = {
+    {DatabaseKind::LevelDB, "leveldb"},
+    {DatabaseKind::MemoryDB, "memorydb"},
+};
+
+void setDatabaseKind(const std::string& _name)
+{
+    for (auto& entry : dbKindsTable)
+    {
+        if (_name == entry.name)
+        {
+            g_kind = entry.kind;
+            return;
+        }
+    }
+
+    BOOST_THROW_EXCEPTION(std::system_error(std::make_error_code(std::errc::invalid_argument),
```

Comment:
I see this comes from `setVMKind`, but I don't think it's a good idea to throw `system_error` here which is usually thrown from standard library classes working with the OS functions.

In general in C++ it's usually advised to define custom exception types deriving `std::exception`, we also utilize `boost::exception`, and the usual approach is to define in some relevant header exception type like
```
DEV_SIMPLE_EXCEPTION(InvalidDatabaseKind);
```

Then throw it here like
```
BOOST_THROW_EXCEPTION(InvalidDatabaseKind());
```
or if you want to include comment
```
BOOST_THROW_EXCEPTION(InvalidDatabaseKind() << errinfo_comment("invalid database kind supplied: " + _name));
```